src/lmdb.ml

```diff
@@ -383,126 +459,48 @@ module Make (Key : Values.S) (Elt : Values.S) = struct
-          mdb_txn_abort txn ; None
-        | exn -> mdb_txn_abort txn ; raise exn
-
-    let abort { txn ; _ } = raise (Abort txn)
-
-    let stats { txn ; db } =
-      let stats = make mdb_stat in
-      mdb_dbi_stat txn db (addr stats) ;
-      Env.make_stats stats
-
-    let flags { txn ; db } =
-      let flags = alloc mdb_dbi_open_flag in
-      mdb_dbi_flags txn db flags ;
-      !@flags
-
-    let drop ?(delete=false) { txn ; db } =
-      mdb_drop txn db delete
-
-    let get { db ; txn } k =
-      let v = addr @@ make mdb_val in
-      mdb_get txn db (dbval_of_buffer @@ write Key.write k) v ;
-      Elt.read @@ buffer_of_dbval v
-
-    let put ?(flags=PutFlags.none) { db ; txn } k v =
-      mdb_put txn db
-        (dbval_of_buffer @@ write Key.write k)
-        (dbval_of_buffer @@ write Elt.write v)
-        flags
-
-    let append t k v =
-      let flags =
-        if has_dup_flag then PutFlags.append_dup else PutFlags.append
-      in
-      put ~flags t k v
-
-    let remove ?elt { db ; txn } k =
-      match elt with
-        | Some v ->
-          mdb_del txn db
-            (dbval_of_buffer @@ write Key.write k)
-            (dbval_of_buffer @@ write Elt.write v)
-        | None ->
-          mdb_del txn db
-            (dbval_of_buffer @@ write Key.write k)
-            (from_voidp mdb_val null)
-
-    let env { txn ; _ } = mdb_txn_env txn
-
-    let compare_key { db ; txn } :key -> key -> int =
-      fun x y ->
-      mdb_cmp txn db
-        (dbval_of_buffer @@ write Key.write x)
-        (dbval_of_buffer @@ write Key.write y)
-
-    let compare_elt ({ db ; txn } as t) :elt -> elt -> int =
-      if not (Flags.(test dup_sort) @@ flags t) then
-        invalid_arg "Lmdb: elements are only comparable in a dup_sort db";
-      fun x y ->
-      mdb_dcmp txn db
-        (dbval_of_buffer @@ write Elt.write x)
-        (dbval_of_buffer @@ write Elt.write y)
-
-    let compare = compare_key
-
-  end
-
-  let append { db ; env } k v =
-    trivial_txn ~write:true env @@ fun txn -> Txn.append {Txn. db ; txn} k v
-
-  let compare_key {db ; env} x y =
-    trivial_txn ~write:false env @@ fun txn -> Txn.compare_key {Txn. db ; txn} x y
-
-  let compare_elt {db ; env} x y =
-    trivial_txn ~write:false env @@ fun txn -> Txn.compare_elt {Txn. db ; txn} x y
+  let compare_key { db ; env } ?txn x y =
+    Txn.maybe ro ?txn env @@ fun txn ->
+    mdb_cmp txn db
+      (dbval_of_buffer @@ write Key.write x)
+      (dbval_of_buffer @@ write Key.write y)
+
+  let compare_elt { db ; env } ?txn :elt -> elt -> int =
+    if not has_dup_flag then
+      invalid_arg "Lmdb: elements are only comparable in a dup_sort db";
+    fun x y ->
+    Txn.maybe ro ?txn env @@ fun txn ->
+    mdb_dcmp txn db
+      (dbval_of_buffer @@ write Elt.write x)
+      (dbval_of_buffer @@ write Elt.write y)
 
   let compare = compare_key
 
-  module Cursor = struct
+  module Crs = struct
 
     type 'a t = mdb_cursor constraint 'a = [< `Read | `Write ]
```

Comment:
In your scheme, this should be contravariant too.

Reply:
yes, it should indeed!